cmd/keygenerator/main.go

```diff
@@ -97,141 +132,338 @@ func main() {
-		return err
+func saveKeys(validatorKeys []key, walletKeys []key, noSplit bool) error {
+	if len(validatorKeys)+len(walletKeys) == 0 {
+		return fmt.Errorf("internal error: no keys to save")
 	}
 
-	genForBlockSigningSk := signing.NewKeyGenerator(ed25519.NewEd25519())
+	var errFound error
+	if len(validatorKeys) > 0 {
+		err := saveSliceKeys(validatorKeyFileName, validatorKeys, validatorPubKeyConverter, noSplit)
+		if err != nil {
+			errFound = err
+		}
+	}
+	if len(walletKeys) > 0 {
+		err := saveSliceKeys(walletKeyFileName, walletKeys, walletPubKeyConverter, noSplit)
+		if err != nil {
+			errFound = err
+		}
+	}
 
-	return generateAndSave(index, numKeys, walletKeyFileName, genForBlockSigningSk, pubkeyConverter)
+	return errFound
 }
 
-func generateAndSave(index int, numKeys int, baseFilename string, genForBlockSigningSk crypto.KeyGenerator, pubkeyConverter core.PubkeyConverter) error {
+func saveSliceKeys(baseFilename string, keys []key, pubkeyConverter core.PubkeyConverter, noSplit bool) error {
+	var file *os.File
+	var err error
+	for i, k := range keys {
+		shouldCreateFile := !noSplit || i == 0
+		if shouldCreateFile {
+			file, err = generateFile(i, len(keys), baseFilename)
+			if err != nil {
+				return err
+			}
+		}
+
+		err = writeKeyToStream(file, k, pubkeyConverter)
+		if err != nil {
+			return err
+		}
+		if !noSplit {
+			err = file.Close()
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	if noSplit {
+		return file.Close()
+	}
+
+	return nil
+}
+
+func generateFile(index int, numKeys int, baseFilename string) (*os.File, error) {
 	folder, err := generateFolder(index, numKeys)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	filename := filepath.Join(folder, baseFilename)
 	backupFileIfExists(filename)
-
 	err = os.Remove(filename)
 	if err != nil && !os.IsNotExist(err) {
-		return err
+		return nil, err
 	}
 
-	file, err := os.OpenFile(filename, os.O_CREATE|os.O_WRONLY, core.FileModeUserReadWrite)
+	return os.OpenFile(filename, os.O_CREATE|os.O_WRONLY, core.FileModeUserReadWrite)
+}
+
+func generateFolder(index int, numKeys int) (string, error) {
+	absPath, err := os.Getwd()
 	if err != nil {
-		return err
+		return "", err
 	}
 
-	defer func() {
-		_ = file.Close()
-	}()
+	if numKeys > 1 {
+		absPath = filepath.Join(absPath, fmt.Sprintf(keysFolderPattern, index))
+	}
+
+	log.Info("generating files in", "folder", absPath)
 
-	sk, pk, err := generateKeys(genForBlockSigningSk)
+	err = os.MkdirAll(absPath, os.ModePerm)
 	if err != nil {
-		return err
+		return "", err
 	}
 
-	pkString := pubkeyConverter.Encode(pk)
+	return absPath, nil
+}
 
-	return core.SaveSkToPemFile(file, pkString, []byte(hex.EncodeToString(sk)))
+func backupFileIfExists(filename string) {
+	if _, err := os.Stat(filename); err != nil {
+		if os.IsNotExist(err) {
+			return
+		}
+	}
+	//if we reached here the file probably exists, make a timestamped backup
+	_ = os.Rename(filename, filename+"."+fmt.Sprintf("%d", time.Now().Unix()))
 }
+
```

Comment:
Don't forget to remove the commented code

Reply:
done, thanks